presto-main/src/main/java/com/facebook/presto/operator/aggregation/Histogram.java

```diff
@@ -0,0 +1,167 @@
+        Type outputType = new MapType(keyType, valueType);
+        HistogramStateSerializer stateSerializer = new HistogramStateSerializer(keyType);
+        Type intermediateType = stateSerializer.getSerializedType();
+        MethodHandle inputFunction = INPUT_FUNCTION.bindTo(keyType);
+
+        AggregationMetadata metadata = new AggregationMetadata(
+                generateAggregationName(NAME, outputType, inputTypes),
+                createInputParameterMetadata(keyType),
+                inputFunction,
+                null,
+                null,
+                COMBINE_FUNCTION,
+                OUTPUT_FUNCTION,
+                HistogramState.class,
+                stateSerializer,
+                new HistogramStateFactory(),
+                outputType,
+                false);
+
+        GenericAccumulatorFactoryBinder factory = new AccumulatorCompiler().generateAccumulatorFactoryBinder(metadata, classLoader);
+        return new InternalAggregationFunction(NAME, inputTypes, intermediateType, outputType, true, false, factory);
+    }
+
+    private static List<ParameterMetadata> createInputParameterMetadata(Type keyType)
+    {
+        return ImmutableList.of(new ParameterMetadata(STATE),
+                new ParameterMetadata(BLOCK_INPUT_CHANNEL, keyType),
+                new ParameterMetadata(BLOCK_INDEX));
+    }
+
+    public static void input(Type type, HistogramState state, Block key, int position)
+    {
+        TypedHistogram typedHistogram = state.get();
+        if (typedHistogram == null) {
+            typedHistogram = new TypedHistogram(type, EXPECTED_SIZE_FOR_HASHING);
+            state.set(typedHistogram);
+        }
+
+        long startSize = typedHistogram.getEstimatedSize();
+        try {
+            typedHistogram.add(position, key, 1L);
+        }
+        catch (ExceededMemoryLimitException e) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("The result of histogram may not exceed %s", e.getMaxMemory()));
+        }
+        state.addMemoryUsage(typedHistogram.getEstimatedSize() - startSize);
+    }
+
+    public static void combine(HistogramState state, HistogramState otherState)
+    {
+        if (state.get() != null && otherState.get() != null) {
+            TypedHistogram typedHistogram = state.get();
+            long startSize = typedHistogram.getEstimatedSize();
+            try {
+                typedHistogram.addAll(otherState.get());
+            }
+            catch (ExceededMemoryLimitException e) {
+                throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("The result of histogram may not exceed %s", e.getMaxMemory()));
+            }
+            state.addMemoryUsage(typedHistogram.getEstimatedSize() - startSize);
+        }
+        else if (state.get() == null) {
+            state.set(otherState.get());
+        }
+    }
+
+    public static void output(HistogramState state, BlockBuilder out)
```

Comment:
add a `Type` parameter to this method and bind it to `outputType` in `specialize()`. Then instead of calling `writeBytes` and `closeEntry` call `type.writeSlice`